(profile::ccs::file_transfer) add option to install s3daemon

glennmorris opened the pull request:
New class parameters:
(s3daemon) boolean, true to install
(s3daemon_repo_url, s3daemon_repo_rev, s3daemon_repo_directory)
  strings for repo URL and revision, and install location
(s3daemon_env_file, s3daemon_env_url, s3daemon_env_access, s3daemon_env_secret)
  strings for environment file location, and contents

When true: install required yum and pip packages; clone repository; configure and start s3daemon system service.

(templates/ccs/service.epp) new file
(templates/ccs/file_transfer/s3daemon_envfile.epp) new file

glennmorris commented:
This requires:
https://github.com/lsst-it/lsst-puppet-hiera-private/pull/110

jhoblitt commented on site/profile/manifests/ccs/file_transfer.pp:
```diff
@@ -39,8 +55,132 @@
+  Boolean $s3daemon = false,
+  Stdlib::HTTPUrl $s3daemon_repo_url = 'https://github.com/lsst-dm/s3daemon',
+  Optional[String] $s3daemon_repo_rev = undef,
+  String $s3daemon_repo_directory = '/home/ccs-ipa/s3daemon/git',
```
It is a good idea to use `String[1]` instead of `String` unless there is a need to support `""`.

jhoblitt commented on site/profile/manifests/ccs/file_transfer.pp:
```diff
@@ -39,8 +55,132 @@
+  String $s3daemon_env_file = '/home/ccs-ipa/s3daemon/env',
+  String $s3daemon_env_url = 'https://s3dfrgw.slac.stanford.edu',
+  String $s3daemon_env_access = 'access-key',
+  String $s3daemon_env_secret = 'secret-key',
```
I assume these are s3 credentials?  Shouldn't these be mandatory parameters?

glennmorris replied:
I suppose that would be clearer, yes.

jhoblitt commented on site/profile/manifests/ccs/file_transfer.pp:
```diff
@@ -39,8 +55,132 @@
+  String $s3daemon_env_secret = 'secret-key',
 ) {
   if $install {
+    if $s3daemon {
```
The nested `if` here is a bit ugly.  Can this new functionality be in a new profile or all of this moved to a stand alone module?  (also, why does the `install` param exist?)

glennmorris replied:
Probably the install parameter exists because I wanted to reuse some parameters from profile::ccs::common, and thought that meant I should include profile::ccs::file_transfer in common, rather than having it be entirely separate. I then needed a way to control which hosts it was applied on. But I suppose it is reasonable to assume that profile::ccs::common will always be in use, so I could indeed make this separate and get rid of install.

In the short/medium term, I believe s3daemon will replace the old way of doing file transfer, so hopefully it is only a temporary situation to need two separate flags.

glennmorris committed:
(profile::ccs::file_transfer) remove the install parameter

Rather than including the class in ccs::common and conditionally
activating it, only include it in the roles that need it.

glennmorris committed:
(profile::ccs::file_transfer) add option to install s3daemon

New class parameters:
(s3daemon) boolean, true to install
(s3daemon_repo_url, s3daemon_repo_rev, s3daemon_repo_directory)
strings for repo URL and revision, and install location
(s3daemon_env_file, s3daemon_env_url, s3daemon_env_access, s3daemon_env_secret)
strings for environment file location, and contents

When true: install required yum and pip packages; clone repository;
configure and start s3daemon system service.

(templates/ccs/service.epp) new file
(templates/ccs/file_transfer/s3daemon_envfile.epp) new file

glennmorris committed:
(spec/fixtures/hieradata/common) add ccs file-transfer s3daemon

glennmorris committed:
(spec/classes/ccs/file_transfer) test s3daemon

glennmorris commented:
@jhoblitt thanks for the comments. I have pushed a new version - please re-review when you have time, thanks.

jhoblitt reviewed (approved):
LGTM.  Thank you for implementing my comments.
